controlplane/kubeadm/internal/workload_cluster.go

```diff
@@ -123,37 +152,108 @@ func (w *Workload) ControlPlaneIsHealthy(ctx context.Context) (HealthCheckResult
+		if _, err = w.checkPodStatusAndUpdateCondition(KubeAPIServerPodNamePrefix, node, owningMachine, clusterv1.MachineAPIServerPodHealthyCondition); err != nil {
+			anyError = err
 		}
-		apiServerPod := corev1.Pod{}
-		if err := w.Client.Get(ctx, apiServerPodKey, &apiServerPod); err != nil {
-			response[name] = err
-			continue
+
+		// Check kube-controller-manager health
+		if _, err = w.checkPodStatusAndUpdateCondition(KubeControllerManagerPodNamePrefix, node, owningMachine, clusterv1.MachineControllerManagerHealthyCondition); err != nil {
+			anyError = err
 		}
-		response[name] = checkStaticPodReadyCondition(apiServerPod)
 
-		controllerManagerPodKey := ctrlclient.ObjectKey{
-			Namespace: metav1.NamespaceSystem,
-			Name:      staticPodName("kube-controller-manager", name),
+		// Check kube-scheduler health
+		if _, err = w.checkPodStatusAndUpdateCondition(KubeSchedulerHealthyPodNamePrefix, node, owningMachine, clusterv1.MachineSchedulerPodHealthyCondition); err != nil {
+			anyError = err
 		}
-		controllerManagerPod := corev1.Pod{}
-		if err := w.Client.Get(ctx, controllerManagerPodKey, &controllerManagerPod); err != nil {
+
+		if anyError != nil {
 			response[name] = err
 			continue
 		}
-		response[name] = checkStaticPodReadyCondition(controllerManagerPod)
 	}
 
 	return response, nil
 }
 
+// checkPodStatusAndUpdateConditions returns true, only if the Pod is in Running/Ready condition.
+// There is a slight difference between returning false and error:
+// checkPodStatusAndUpdateConditions returns false, if Pod is not in Ready Condition or there is a client error.
+// checkPodStatusAndUpdateConditions returns error only when it is obvious that the Pod is not in Running/Ready state, so it does not return error on transient client errors.
+// For instance, returning (false, nil) means Pod is not ready but is not in error state as well,
+// returning (false, err) means Pod is not ready and it is in error state as well.
+func (w *Workload) checkPodStatusAndUpdateCondition(staticPodPrefix string, node corev1.Node, owningMachine *clusterv1.Machine, conditionType clusterv1.ConditionType) (bool, error) {
+	staticPodKey := ctrlclient.ObjectKey{
+		Namespace: metav1.NamespaceSystem,
+		Name:      staticPodName(staticPodPrefix, node.Name),
+	}
+
+	staticPod := corev1.Pod{}
+	if err := w.Client.Get(context.TODO(), staticPodKey, &staticPod); err != nil {
+		// If there is an error getting the Pod, do not set any conditions.
+		if apierrors.IsNotFound(err) {
+			if owningMachine != nil {
+				conditions.MarkFalse(owningMachine, conditionType, clusterv1.PodMissingReason, ConditionReason(clusterv1.PodMissingReason).GetSeverity(), "")
+			}
+			return false, errors.Errorf("static pod %s is missing", staticPodPrefix)
+		}
+		// Do not return error here, because this may be a transient error.
+		return false, nil
+	}
+
+	if err := checkStaticPodReadyCondition(staticPod); err != nil {
+		if checkStaticPodFailedPhase(staticPod) {
```

Comment:
Flip this, as in check `err == nil` first, which has the least amount of code in the block, and then check the rest afterwards.

Sidenote, I've been seeing a lot of `checkX` functions, which are one-off checks. Let's try to see if we can identify a pattern that avoids generic checks, but it's rather more focused

Reply:
I don't want to add all logic into the same function, some abstraction like `isStaticPodInReadyCondition`, `isStaticPodInFailedPhase`, `isStaticPodProvisioning` makes it easier to read here IMO.